Dynamic point following tutorial (#69)
ros-planning/navigation.ros.org: tutorials/docs/navigation2_dynamic_point_following.rst

.. _navigation2-dynamic-point-following:

> Dynamic point following

Reviewer: ```suggestion
Dynamic Object Following
```
Capitalize titles, and point isn't really what we're doing, its a detected object. That makes things more clear

> ***********************

Reviewer: The header lengths must be the same as the header letter count.

Author: fixed

- `Overview`_
- `Tutorial Steps`_

.. raw:: html

    <h1 align="center">
      <div style="position: relative; padding-bottom: 0%; overflow: hidden; max-width: 100%; height: auto;">
        <iframe width="700" height="450" src="https://www.youtube.com/embed/sRodzrrJChA?autoplay=1" frameborder="1" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
      </div>
    </h1>

Overview
========

> This tutorial shows how to use Navigation2 for a different task of going from point A to point B. In this case, we will use Navigation2 to follow a moving object at a distance indefinitely.

Reviewer: ```suggestion
This tutorial shows how to use Navigation2 for a different task other than going from point A to point B. In this case, we will use Navigation2 to follow a moving object at a distance indefinitely.
```

This task is useful in cases such as following a person, as shown in the next video of "Carry My Luggage" RoboCup @ Home test, in which the `CATIE Robotics <https://robotics.catie.fr/>`_ team performs the test successfully, or this other video 
of a real (future) world application:

.. raw:: html

    <h1 align="center">
      <div>
        <div style="position: relative; padding-bottom: 0%; overflow: hidden; max-width: 100%; height: auto;">
          <iframe width="450" height="300" src="https://www.youtube.com/embed/lTjKO4M7yZc?autoplay=1&mute=1" frameborder="1" allowfullscreen></iframe>
          <iframe width="450" height="300" src="https://www.youtube.com/embed/KgRKyzsja9Q?autoplay=1&mute=1" frameborder="1" allowfullscreen></iframe>
        </div>
      </div>
    </h1>

The requirements for this task are as follows:

> - The changes to achieve this functionality are limited to the behavior tree used to navigate. This behavior tree can be selected in the NavigateToPose action when this functionality is required, or it can be the default behavior tree.

Reviewer: ```suggestion
- The changes are limited to the behavior tree used to navigate. This behavior tree can be selected in the ``NavigateToPose`` action when required, or it can be the default behavior tree.
```

send the initial pose to the ``NavigateToPose`` action, and update it on a topic for the duration of the task:

.. image:: images/navigation2_dynamic_point_following/main_diagram.png
    :width: 48%

Before completing this tutorial, please look at the previous two tutorials on navigation in simulation and physical hardware.
This tutorial assumes knowledge of navigation and basic understanding of behavior trees.

Tutorial Steps
==============

0- Create the Behavior Tree
---------------------------

Let's start from this simple behavior tree. This behavior tree with replan a new path every 1hz and pass that path to the controller to follow:

.. code-block:: xml

  <root main_tree_to_execute="MainTree">
    <BehaviorTree ID="MainTree">
      <PipelineSequence name="NavigateWithReplanning">
        <RateController hz="1.0">
          <ComputePathToPose goal="{goal}" path="{path}" planner_id="GridBased"/>
        </RateController>
        <FollowPath path="{path}" controller_id="FollowPath"/>
      </PipelineSequence>
    </BehaviorTree>
  </root>

> First of all, let's make that this behavior runs while there is not any failure. For this purpose, we will use the ``KeepRunningUntilFailure`` control node.

Reviewer: ```suggestion
First of all, let's make this behavior run until there's a failure. For this purpose, we will use the ``KeepRunningUntilFailure`` control node.
```

      </PipelineSequence>
    </BehaviorTree>
  </root>

To stay at a certain distance from the target, we will use the action node ``TruncatePath``. This node modifies a path making it shorter so we don't try to navigate into the object of interest. We can set up the desired distance to the goal using the input port ``distance``.

.. code-block:: xml

  <root main_tree_to_execute="MainTree">
    <BehaviorTree ID="MainTree">
      <PipelineSequence name="NavigateWithReplanning">
        <RateController hz="1.0">
          <Sequence>
            <GoalUpdater input_goal="{goal}" output_goal="{updated_goal}">
              <ComputePathToPose goal="{updated_goal}" path="{path}" planner_id="GridBased"/>
            </GoalUpdater>
           <TruncatePath distance="1.0" input_path="{path}" output_path="{truncated_path}"/>
          </Sequence>
        </RateController>
        <KeepRunningUntilFailure>
          <FollowPath path="{truncated_path}" controller_id="FollowPath"/>
        </KeepRunningUntilFailure>
      </PipelineSequence>
    </BehaviorTree>
  </root>

Let's save this behavior tree and use it in our navigation task.

1- Setup Rviz clicked point
---------------------------

> In this tutorial, we are going to use RViz instead of a full application. We will use the "clicked point" button on the toolbar to send goal updates to Navigation2. This button allows you to 

Reviewer: ```suggestion
In this tutorial, we are going to use RViz instead of a full application. We will use the "clicked point" button on the toolbar to substitute object detections to provide goal updates to Navigation2. This button allows you to 
```

publish coordinates in the topic ``/clicked_point``. This point needs to be sent to the behavior tree, using the program ``clicked_point_to_pose``, from `this repo <https://github.com/fmrico/nav2_test_utils>`_. Clone 
this repo in your workspace, build, and type in a terminal.

``ros2 run nav2_test_utils clicked_point_to_pose``

2- Run Dynamic Object Following in Navigation2 Simulation
---------------------------------------------------------

Start Navigation2 in one terminal:

``ros2 launch nav2_bringup tb3_simulation_launch.py default_bt_xml_filename:=/path/to/bt.xml``

> Open RViz and, after initialize the robot position, command the robot to navigate to any position. Use the button clicked point to change the goal, as shown in the video in the head of this tutorial.

Reviewer: ```suggestion
Open RViz and, after initialize the robot position, command the robot to navigate to any position. Use the button clicked point to simulate a new detection of the object of interest, as shown in the video in the head of this tutorial.
```

.. _navigation2-dynamic-point-following:

> Dynamic point following to a certain distance

Reviewer: "Dynamic Object Following" I think is a good substitute. A little shorter (so that it all appears in 1 line for peoples' browsers) and I think the distance part is implied and later explained in the document

Author: changed

*********************************************

- `Overview`_
- `Tutorial Steps`_

.. raw:: html

>     <h1 align="center">

Reviewer: This video is not super clear as to what's going on or how this is showing off this capability. Overall, I think this should just be removed since I don't think there's a way to clearly demonstrate this capability without a physical robot physically following something. Else, it just appears as a waypoint follower which isn't as compelling.

Reviewer: I think the 2nd video showing a physical robot should replace this video. I'm not sure if you have access to robots due to COVID to generate a better video where the person moves faster (and speed 5x for brevity)

Author: Ok. I will remove it and try to select the videos that better illustrate the page.

Now is complicated to do a video with the real robot.

      <div style="position: relative; padding-bottom: 0%; overflow: hidden; max-width: 100%; height: auto;">
        <iframe width="700" height="450" src="https://www.youtube.com/embed/rUCM71TWtGM?autoplay=1" frameborder="1" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
      </div>
    </h1>

Overview
========

This tutorial shows how to use Navigation2 for a different task of going from point A to point B. In this case, we will use Navigation2 to follow a moving object at a distance indefinitely.

> This task is useful in cases such as following a person, as shown in this video of "Follow me" RoboCup@Home test, in which the `CATIE Robotics <https://robotics.catie.fr/>`_ team performs the test successfully:

Reviewer: The "RoboCup@Home" renders as a hyperlink to a URL, might be worth adding some spaces "RoboCup @ Home" to help stop that. 

I like the link to the team, do they have a video of this on their robot?

It might be good to show a non-research application link as well like https://www.youtube.com/watch?v=KgRKyzsja9Q for the suitcases.

Author: An alternative video of this team is https://www.youtube.com/watch?v=lTjKO4M7yZc , but I think that this video could show better the behavior.

Ok, I will add the video that you suggest

.. raw:: html

    <h1 align="center">
      <div style="position: relative; padding-bottom: 0%; overflow: hidden; max-width: 100%; height: auto;">
        <iframe width="700" height="450" src="https://www.youtube.com/embed/sRodzrrJChA?autoplay=1" frameborder="1" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
      </div>
    </h1>

The requirements for this task are as follows:

- The changes to achieve this functionality are limited to the behavior tree used to navigate. This behavior tree can be selected in the NavigateToPose action when this functionality is required, or it can be the default behavior tree.
- The configuration of the planner and the controller will not be modified.
- The action will indefinitely run until it is canceled by who initiated it.

> The calculation of the dynamic pose to follow is outside the scope of this tutorial. As shown in the following diagram, the application using Navigation2 is supposed to calculate this position, 

Reviewer: ```suggestion
The detection of the dynamic object (person) pose to follow is outside the scope of this tutorial. As shown in the following diagram, your application should provide a detector for the object(s) of interest, 
```

> send it in the call to the NavigateToPose action, and update it during the execution of this task:

Reviewer: ```suggestion
send the initial pose to the ``NavigateToPose`` action, and update it on a topic for the duration of the task:
```

> .. image:: images/navigation2_dynamic_point_following/main_diagram.png

Reviewer: Thanks for adding a diagram!

Though I'm not sure what "behavior" means in the top left block. I'd also specify the type on the updated pose arrow like you do on the navigate to pose one & specify that its going to the specific BT node, not to the behavior tree abstractly.

Author: changing...

    :width: 48%

> Before completing this tutorial, please look at the previous two tutorials on navigation in simulation and physical hardware, if available.

Reviewer: ```suggestion
Before completing this tutorial, please look at the previous two tutorials on navigation in simulation and physical hardware.
```

This tutorial assumes knowledge of navigation and basic understanding of behavior trees.

Tutorial Steps
==============

0- Create the Behavior Tree
---------------------------

> Let's start from this simple behavior tree:

Reviewer: Please  explain what they're looking at. Something short like "This behavior tree with replan a new path every 1hz and pass that path to the controller to follow."

Author: done

.. code-block:: xml

  <root main_tree_to_execute="MainTree">
    <BehaviorTree ID="MainTree">
      <PipelineSequence name="NavigateWithReplanning">
        <RateController hz="1.0">
          <ComputePathToPose goal="{goal}" path="{path}" planner_id="GridBased"/>
        </RateController>
        <FollowPath path="{path}" controller_id="FollowPath"/>
      </PipelineSequence>
    </BehaviorTree>
  </root>

First of all, let's make that this behavior runs while there is not any failure. For this purpose, we will use the ``KeepRunningUntilFailure`` control node.

.. code-block:: xml

  <root main_tree_to_execute="MainTree">
    <BehaviorTree ID="MainTree">
      <PipelineSequence name="NavigateWithReplanning">
        <RateController hz="1.0">
          <ComputePathToPose goal="{goal}" path="{path}" planner_id="GridBased"/>
        </RateController>
        <KeepRunningUntilFailure>
          <FollowPath path="{path}" controller_id="FollowPath"/>
        </KeepRunningUntilFailure>
      </PipelineSequence>
    </BehaviorTree>
  </root>

> We will use the decorator GoalUpdater to use a dynamic point. This node takes as input the requested goal in the ``NavigateToPose`` action and subscribes to the topic ``/goal_update``. It set as output ``updated_goal``:

Reviewer: ```suggestion
We will use the decorator ``GoalUpdater`` to update the dynamic object pose to follow. This node takes as input the current goal and subscribes to the topic ``/goal_update``. It set the new goal as ``updated_goal``:
```

> - The original goal, if nothing received in ``/goal_update``.

Reviewer: I think these are supposed to be bullets, you need to add a space between each one so they dont all appear on one line. See the rendering https://337-240847415-gh.circle-artifacts.com/0/html/tutorials/docs/navigation2_dynamic_point_following.html

- Or, the pose received in ``/goal_update``.

.. code-block:: xml

  <root main_tree_to_execute="MainTree">
    <BehaviorTree ID="MainTree">
      <PipelineSequence name="NavigateWithReplanning">
        <RateController hz="1.0">
          <GoalUpdater input_goal="{goal}" output_goal="{updated_goal}">
            <ComputePathToPose goal="{updated_goal}" path="{path}" planner_id="GridBased"/>
          </GoalUpdater>
        </RateController>
        <KeepRunningUntilFailure>
          <FollowPath path="{path}" controller_id="FollowPath"/>
        </KeepRunningUntilFailure>
      </PipelineSequence>
    </BehaviorTree>
  </root>

> To stay at a certain distance from the target, we will use the action node ``TruncatePath``. This node modifies a path making it shorter. We can set up the desired distance to the goal using the input port ``distance``.

Reviewer: ```suggestion
To stay at a certain distance from the target, we will use the action node ``TruncatePath``. This node modifies a path making it shorter so we don't try to navigate into the object of interest. We can set up the desired distance to the goal using the input port ``distance``.
```

.. code-block:: xml

  <root main_tree_to_execute="MainTree">
    <BehaviorTree ID="MainTree">
      <PipelineSequence name="NavigateWithReplanning">
        <RateController hz="1.0">
          <Sequence>
            <GoalUpdater input_goal="{goal}" output_goal="{updated_goal}">
              <ComputePathToPose goal="{updated_goal}" path="{path}" planner_id="GridBased"/>
            </GoalUpdater>
           <TruncatePath distance="1.0" input_path="{path}" output_path="{truncated_path}"/>
          </Sequence>
        </RateController>
        <KeepRunningUntilFailure>
          <FollowPath path="{truncated_path}" controller_id="FollowPath"/>
        </KeepRunningUntilFailure>
      </PipelineSequence>
    </BehaviorTree>
  </root>

> Let's save this behavior tree in ``nav2_bt_navigator/behavior_trees/follow_point.xml``

Reviewer: ```suggestion
Let's save this behavior tree and use it in our navigation task
```

Reviewer: That file exists in the repo, so we dont want people to override it

2- Modify launcher
------------------

Select the above behavior tree as the default in the launcher ``tb3_simulation_launch.py``:

.. code-block:: python

>   declare_bt_xml_cmd = DeclareLaunchArgument(

Reviewer: Rather than this, why not have them set the command argument? So, something along the lines of: rather than running ``tb3_simulation_launch.py`` with no launch arguments, we will set the behavior tree to follow dynamic points. Then in the code block, have the `ros2 launch ... default_bt_xml_filename:=/path/to/bt.xml`

Author: ok

      'default_bt_xml_filename',
      default_value=os.path.join(
          get_package_share_directory('nav2_bt_navigator'),
          'behavior_trees', 'follow_point.xml'),
      description='Full path to the behavior tree xml file to use')

3- Setup Rviz clicked point
---------------------------

In this tutorial, we are going to use RViz instead of a full application. We will use the "clicked point" button on the toolbar to send goal updates to Navigation2. This button allows you to 
publish coordinates in the topic ``/clicked_point``. This point needs to be sent to the behavior tree, using the program ``clicked_point_to_pose``, from `this repo <https://github.com/fmrico/nav2_test_utils>`_. Clone 
this repo in your workspace, build, and type in a terminal.

``ros2 run nav2_test_utils clicked_point_to_pose``

> 4- Run Navigation2

Reviewer: ```suggestion
4- Run Dynamic Object Following in Navigation2 Simulation
```

Author: ok
